### PR TITLE
[RHCLOUD-22065] Feature store webhook secrets sources

### PR DESCRIPTION
**Pull request by MikelAlejoBR**
Enables storing the basic authentication and secret token secrets in Sources. It can get turned on and off by the `notifications.use-sources-secrets-backend` feature flag.

 ## Links

[[RHCLOUD-22065]](https://issues.redhat.com/browse/RHCLOUD-22065)

### CODE REVIEW

File: database/src/main/resources/db/migration/V1.65.0__NOTIF-619_webhook_sources_auths.sql
Comment (josejulio) [inline database/src/main/resources/db/migration/V1.65.0__NOTIF-619_webhook_sources_auths.sql:12]: Could you merge these 2 in a single `ALTER` ?

```suggestion
ALTER TABLE
  endpoint_webhooks
ADD COLUMN
    basic_authentication_id BIGINT,
ADD COLUMN
    secret_token_id BIGINT
```

File: database/src/main/resources/db/migration/V1.65.0__NOTIF-619_webhook_sources_auths.sql
Comment (josejulio) [inline database/src/main/resources/db/migration/V1.65.0__NOTIF-619_webhook_sources_auths.sql:25]: Same as above
```suggestion
ALTER TABLE
    camel_properties
ADD COLUMN
    basic_authentication_id BIGINT,
ADD COLUMN
    secret_token_id BIGINT
;
```

File: database/src/main/resources/db/migration/V1.65.0__NOTIF-619_webhook_sources_auths.sql
Comment (MikelAlejoBR) [reply database/src/main/resources/db/migration/V1.65.0__NOTIF-619_webhook_sources_auths.sql]: Done, thanks!

File: database/src/main/resources/db/migration/V1.65.0__NOTIF-619_webhook_sources_auths.sql
Comment (MikelAlejoBR) [reply database/src/main/resources/db/migration/V1.65.0__NOTIF-619_webhook_sources_auths.sql]: Done as well, thanks!

File: common/src/main/java/com/redhat/cloud/notifications/models/WebhookProperties.java
Comment (josejulio) [inline common/src/main/java/com/redhat/cloud/notifications/models/WebhookProperties.java:43]: Lets use the Object-type equivalent as we could have `null` values in the database - unless the migrations sets a default value.

```suggestion
    private Long secretTokenSourcesId;
```

File: common/src/main/java/com/redhat/cloud/notifications/models/WebhookProperties.java
Comment (josejulio) [inline common/src/main/java/com/redhat/cloud/notifications/models/WebhookProperties.java:55]: Same as above - we will need to do the same for the other fields that we are adding

```suggestion
    private Long basicAuthenticationSourcesId;
```

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java:45]: FYI, there is a newer set of annotations available in Quarkus that is a bit more dev-friendly:
```suggestion
    Secret getById(@RestPath long id);
```

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java:21]: Links to https://console.redhat.com/docs/api/sources/v3.1 or the equivalent for stage could be helpful here. We can still find the raw JSON files from there.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java:131]: Nitpick:
```suggestion
     * Deletes the endpoint's secrets. It requires for the properties to have a "basic authentication" ID or "secret
     * token" ID on the database.
```

File: backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java:473]: Unless I missed something, this could throw a `ClassCastException` if `endpoint` does not implement `SourcesSecretable`. It would be better to check if `endpoint instanceof SourcesSecretable` before doing this.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java:478]: What if `endpoint` did not contain any basic auth or secret token data when it was first created, but now it contains that data? I suppose we'll need to create data in Sources.

What if `endpoint` previously contained basic auth or secret token data but they were removed by the user and now everything's empty/null? I suppose we'll need to remove the old data from Sources.

File: common/src/main/java/com/redhat/cloud/notifications/models/WebhookProperties.java
Comment (MikelAlejoBR) [reply common/src/main/java/com/redhat/cloud/notifications/models/WebhookProperties.java]: Yeah, makes sense!

File: common/src/main/java/com/redhat/cloud/notifications/models/WebhookProperties.java
Comment (MikelAlejoBR) [reply common/src/main/java/com/redhat/cloud/notifications/models/WebhookProperties.java]: Done!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/Secret.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/sources/Secret.java:14]: We usually add `@JsonIgnoreProperties(ignoreUnknown = true)` to all classes that are used to deserialize incoming REST data. It prevents deserialization failures if the remote API adds new fields without letting us know. You know the Sources API way better than me so I'll let you decide if we need it or not. 😄

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: I like the `load` naming better too. Changed!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java]: Done!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesXRHIDHeaderFactory.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesXRHIDHeaderFactory.java:29]: The idea here was to simply forward the `x-rh-identity` header, since that's the only one that Sources requires for the request to be authenticated. By adding all the `outgoingHeaders`, wouldn't we be sending also headers that might be specific for Notifications too? :thinking:

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java]: Nice, I like that one better!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java:20]: I think I included links to the stage, production and then the GitHub repository directly. Maybe I missed something?

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java:131]: Fixed! Nice catch :fishing_pole_and_fish:

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/Secret.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/Secret.java:16]: It shouldn't happen, but I rather include it and be safe than sorry :+1:

File: backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java]: Nice catch :fishing_pole_and_fish: ! I was just trying with `Webhook` and `Camel` endpoints and I forgot about that type check

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java:20]: Sorry, my comment wasn't clear enough. I meant that a link to https://console.redhat.com/docs/api/sources/v3.1 is more helpful (from my perspective) than a link to https://console.redhat.com/api/sources/v3.1/openapi.json because of the user-friendly UI 😃

Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java]: Oooooh sure, my bad! Yeah, I agree the Swagger UI is just better. Changing it!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesXRHIDHeaderFactory.java
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesXRHIDHeaderFactory.java:29]: `outgoingHeaders` is currently empty because the `SourcesService` methods are not annotated with `@ClientHeaderParam` or `@HeaderParam`. This will probably never change and your code will most likely work in the future 👍

However, if this class is copied (which happens a lot) to build something else, then the copied class may remove the outgoing headers unintentionally and the dev may waste time understanding why this is happening. So I would merge the outgoing headers here (it won't hurt) even if it's not needed for now 😃

Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesXRHIDHeaderFactory.java]: Ah, I see. Totally understandable. I included the suggestion too! :+1:

File: backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java]: > What if endpoint did not contain any basic auth or secret token data when it was first created, but now it contains that data? I suppose we'll need to create data in Sources.

Yeah, agree on this one.

> What if endpoint previously contained basic auth or secret token data but they were removed by the user and now everything's empty/null? I suppose we'll need to remove the old data from Sources.

Good question. I'm guessing that since the clients have to send the entire object to our `PUT /endpoints/{id}`, it is safe for us to assume that when the basic authentication or the secret token are `null`, we can safely delete the secrets?

Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/EndpointResource.java]: All right, I did implement this behaviour too for the "update" operation!

File: backend/src/main/resources/application.properties
Comment (MikelAlejoBR) [reply backend/src/main/resources/application.properties]: Yeah, it should be `clowder.endpoints.sources-api`. Fixed that, thanks!

**Comment by gwenneg** [comment]
@MikelAlejoBR Note for later: `notifications-engine` will need an update to consume the encrypted credentials stored in Sources. I'm thinking of `CamelTypeProcessor` and `WebhookTypeProcessor` but there may be more places to update. Considering the changes from this PR are disabled by the feature flag, I think it's safe to merge this one and then create another PR for the second part of the work.

File: common/src/main/java/com/redhat/cloud/notifications/models/BasicAuthentication.java
Comment (MikelAlejoBR) [reply common/src/main/java/com/redhat/cloud/notifications/models/BasicAuthentication.java:7]: Do I rollback these changes then?

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java:36]: I added the fault tolerance bit so that we can be sure that failures are not due to "one-off" errors. I added the extension by running the `quarkus extension add 'smallrye-fault-tolerance'` command. That seemed to stir the `pom.xml` file a bit. Is this okay or do you usually add extension manually?

**Comment by MikelAlejoBR** [comment]
Sorry about the force push. Basically I reworded all the commits to replace the `NOTIF-` ticket with the `RHCLOUD-` one.

File: common/src/main/java/com/redhat/cloud/notifications/models/BasicAuthentication.java
Comment (gwenneg) [reply common/src/main/java/com/redhat/cloud/notifications/models/BasicAuthentication.java:7]: Yeah. These changes are most likely good but let's do that another time.

Comment (MikelAlejoBR) [reply common/src/main/java/com/redhat/cloud/notifications/models/BasicAuthentication.java]: All right! Rolled back.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java:35]: Done as well! I've created a ticket for the CLI stuff: [RHCLOUD-22602](https://issues.redhat.com/browse/RHCLOUD-22602)

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java:109]: `Then we` what? 😄

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java:111]: Should we delete the data from Sources if `basicAuth` is not null but contains null/empty fields?

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java:65]: `basicAuth` could contains null/empty fields. We should probably not create anything in Sources if there's no actual data to put there.

File: pom.xml
Comment (MikelAlejoBR) [reply pom.xml]: All right, done!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java]: Sure, that'll improve the flexibility!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SourcesService.java]: Done!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: You are right, yes. I guess this comes from when I was using primitive types and these might be zero. But in this cases if we don't have a value for the IDs, the `Long` variables will always be null, so... I removed the redundant condition.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: Good catch, yes. Do you think that we maybe should do the same for the "basic authentication"? If the username *and* password are blank, simply ignore the creation/update.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: Oops! :eyes:

Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: Actually I've removed the comment since this was already explained in the Javadoc of the method.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: Looking at the other review comments the answer is "yes". I'm preparing something for that.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java:107]: Yes, working on it...

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: Done, thanks!

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: Solved!

File: backend/src/test/java/com/redhat/cloud/notifications/routers/sources/SecretUtilsTest.java
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/routers/sources/SecretUtilsTest.java:399]: You could use this here and at a few other places in this class:
```suggestion
        Mockito.verifyNoInteractions(sourcesServiceMock);
```

File: backend/pom.xml
Comment (gwenneg) [inline backend/pom.xml:183]: Nitpick: Could you please move this up to the `<!-- Quarkus -->` "section" of this file?

File: backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java:74]: This needs to be fixed before the merge.

Comment (MikelAlejoBR) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/sources/SecretUtils.java]: Yes, I forgot to commit it. Done!

File: backend/src/test/java/com/redhat/cloud/notifications/routers/sources/SecretUtilsTest.java
Comment (MikelAlejoBR) [reply backend/src/test/java/com/redhat/cloud/notifications/routers/sources/SecretUtilsTest.java]: Much better. Refactored!

File: backend/pom.xml
Comment (MikelAlejoBR) [reply backend/pom.xml]: Moved!

**Review by gwenneg — APPROVED** [review]
LGTM

**Comment by gwenneg** [comment]
PR check failures are unrelated.

**Review by gwenneg — APPROVED** [review]

**Pull request merged**
